google/cloud/aiplatform/datasets/datasources.py

```diff
@@ -0,0 +1,140 @@
+import abc
+from typing import Optional, Dict, Sequence
+from google.cloud.aiplatform_v1beta1 import (
+    GcsSource,
+    ImportDataConfig,
+)
+from google.cloud.aiplatform import utils
+
+
+class _Datasource(abc.ABC):
+    """An abstract class that sets dataset_metadata"""
+
+    @property
+    @abc.abstractmethod
+    def dataset_metadata(self):
+        """Dataset Metadata."""
+        pass
+
+
+class _DatasourceImportable(abc.ABC):
+    """An abstract class that sets import_data_config"""
+
+    @property
+    @abc.abstractmethod
+    def import_data_config(self):
+        """Import Data Config."""
+        pass
+
+
+class TabularDatasource(_Datasource):
+    """Datasource for creating a tabular dataset for AI Platform"""
+
+    def __init__(
+        self,
+        gcs_source: Optional[Sequence[str]] = None,
+        bq_source: Optional[str] = None,
+    ):
+        """Creates a tabular datasource
+
+        Args:
+            gcs_source: Optional[Sequence[str]]=None:
+                Cloud Storage URI of one or more files. Only CSV files are supported.
+                The first line of the CSV file is used as the header.
+                If there are multiple files, the header is the first line of
+                the lexicographically first file, the other files must either
+                contain the exact same header or omit the header.
+            bq_source: Optional[str]=None:
+                The URI of a BigQuery table.
+
+        Raises:
+            ValueError if source configuration is not valid.
+        """
+
+        dataset_metadata = None
+
+        if gcs_source:
+            gcs_source = utils.sequence_to_list(gcs_source)
+
+        if gcs_source and bq_source:
+            raise ValueError("Only one of gcs_source or bq_source can be set.")
+
+        if not any([gcs_source, bq_source]):
+            raise ValueError("One of gcs_source or bq_source must be set.")
+
+        if gcs_source:
+            dataset_metadata = {"input_config": {"gcs_source": {"uri": gcs_source}}}
+        elif bq_source:
+            dataset_metadata = {"input_config": {"bigquery_source": {"uri": bq_source}}}
+
+        self._dataset_metadata = dataset_metadata
+
+    @property
+    def dataset_metadata(self) -> Optional[Dict]:
+        """Dataset Metadata."""
+        return self._dataset_metadata
+
+
+class NonTabularDatasource(_Datasource):
+    """Datasource for creating a non-tabular dataset for AI Platform"""
+
+    def __init__(self):
+        """Creates an empty non-tabular datasource"""
+        self._dataset_metadata = {}
+
+    @property
+    def dataset_metadata(self) -> Optional[Dict]:
+        return self._dataset_metadata
```

Comment:
Could probably just return `{}` here instead of saving to `self._dataset_metadata`

Reply:
Better yet, just return `None`